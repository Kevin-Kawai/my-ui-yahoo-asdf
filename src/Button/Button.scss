@use "../styles/variables.scss";

.buiButtonBase {
  border-radius: 0.5rem;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  --bui-button-padding-left: var(--bui-spacing-10);
  --bui-button-padding-right: var(--bui-spacing-10);
  padding: 0 var(--bui-spacing-10);
  cursor: pointer;
  height: 4.5rem;
  white-space: nowrap;
  line-height: 1;
  --bui-button-main-color: var(--bui-colors-blue-400);
  --bui-button-sub-color: #fff;
  --bui-button-disabled-color: var(--bui-colors-gray-500);
  // aタグでも使えるように明示的に指定
  box-sizing: border-box;
}
.buiButtonBase,
.buiButtonBase:hover {
  // hoverでもCSSのルールが明示されていないと、 global.cssのhoverが優先されてしまうため
  color: var(--_bui-button-font-color);
  fill: var(--_bui-button-font-color);
  background-color: var(--_bui-button-background-color);
  border: 1px solid var(--_bui-button-border-color);
}

.buiButtonBase:disabled {
  cursor: not-allowed;
}
.buiButtonHoverVariantTransparent:hover,
.buiButtonVariantOutline:hover {
  opacity: 0.6;
}
.buiButtonVariantFilled,
.buiButtonHoverVariantFilled:hover {
  --_bui-button-background-color: var(--bui-button-main-color);
  --_bui-button-font-color: var(--bui-button-sub-color);
  --_bui-button-border-color: var(--bui-button-main-color);
  opacity: revert;
}
.buiButtonVariantFilled:disabled {
  --_bui-button-background-color: var(--bui-button-disabled-color);
  --_bui-button-border-color: var(--bui-button-disabled-color);
}

.buiButtonVariantOutline,
.buiButtonHoverVariantOutline:hover {
  --_bui-button-background-color: var(--bui-button-sub-color);
  --_bui-button-font-color: var(--bui-button-main-color);
  --_bui-button-border-color: var(--bui-button-main-color);
  opacity: revert;
  font-weight: bold;
}

.buiButtonVariantOutline:disabled {
  --_bui-button-border-color: var(--bui-button-disabled-color);
  --_bui-button-font-color: var(--bui-colors-gray-500);
}

.buiButtonSizeXs {
  padding: 0.8rem 1.6rem;
  font-size: var(--bui-font-size-xs);
  height: 2.8rem;
}
.buiButtonSizeSm {
  padding: 0 2rem;
  height: 3.6rem;
}
.buiButtonSizeMd {
  height: 4.5rem;
}
.buiButtonSizeLg {
  height: 4.8rem;
}
.buiButtonSizeXl {
  height: 5.6rem;
}
.buiButtonSize2Xl {
  height: 6rem;
}

.buiButtonBlock {
  width: 100%;
}
.buiButtonRound {
  border-radius: 100vmax;
}
.buiButtonFullWidth {
  padding: revert;
  width: 100%;
}
.buiButtonGapSizeSm {
  gap: var(--bui-spacing-1);
}
.buiButtonGapSizeMd {
  gap: var(--bui-spacing-3);
}
.buiButtonGapSizeLg {
  gap: var(--bui-spacing-4);
}

.buiButtonLeftIconLayoutMd {
  padding-left: 2.2rem;
  padding-right: var(--bui-spacing-10);
}
.buiButtonRightIconLayoutMd {
  padding-left: var(--bui-spacing-10);
  padding-right: 2.2rem;
}
