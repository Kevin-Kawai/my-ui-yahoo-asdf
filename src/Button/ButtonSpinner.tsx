import * as React from "react";

export const ButtonSpinner = () => {
  return (
    <svg
      width="24px"
      height="24px"
      viewBox="0 0 24 24"
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>icon / icon-loading</title>
      <desc>Created with Sketch.</desc>
      <g
        id="icon-/-icon-loading"
        stroke="none"
        strokeWidth="1"
        fill="none"
        fillRule="evenodd"
      >
        <circle
          id="Oval"
          strokeOpacity="0.5"
          stroke="#FFFFFF"
          strokeWidth="2.5"
          fillRule="nonzero"
          cx="12"
          cy="12"
          r="9"
        ></circle>
        <path
          d="M18.5932699,18 C20.0866659,16.4068549 21,14.2716749 21,11.924953 C21,6.99583756 16.9705627,3 12,3"
          id="Path"
          strokeOpacity="0.9"
          stroke="#FFFFFF"
          strokeWidth="2.8"
          strokeLinecap="round"
          strokeLinejoin="round"
          fillRule="nonzero"
        >
          <animateTransform
            attributeType="xml"
            attributeName="transform"
            type="rotate"
            from="0 12 12"
            to="360 12 12"
            dur="0.6s"
            repeatCount="indefinite"
          />
        </path>
      </g>
    </svg>
  );
};
