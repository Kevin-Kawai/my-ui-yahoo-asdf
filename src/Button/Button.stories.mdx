import { Meta, Story, Canvas, ArgsTable } from "@storybook/blocks";
import { Button } from "./Button";
import { TestIcon } from "./TestIcon";

<Meta
  title="boron-ui/Button"
  component={Button}
  parameters={{
    viewMode: "docs",
    previewTabs: {
      canvas: { hidden: true },
    },
  }}
/>

# Button

## Filled(default)

<Canvas>
  <Story name="default">
    <Button>test</Button>
  </Story>
  <Story name="default-disabled">
    <Button disabled>test</Button>
  </Story>
  <Story name="default-loading">
    <Button isLoading>test</Button>
  </Story>
  <Story name="default-sm">
    <Button size="sm">smのButton</Button>
  </Story>
  <Story name="default-large">
    <Button size="lg">lgのButton</Button>
  </Story>
  <Story name="default-xl">
    <Button size="xl">xlのButton</Button>
  </Story>
  <Story name="default-2xl">
    <Button size="2xl">2xlのButton</Button>
  </Story>
  <Story name="default-link">
    <a
      className="buiButtonBase buiButtonVariantFilled"
      href="http://www.google.com/"
      target="_blank"
    >
      これはボタンの見た目をしたリンク
    </a>
  </Story>
</Canvas>

## Outline

<Canvas>
  <Story name="outline">
    <Button variant="outline">test</Button>
  </Story>
  <Story name="outline-disabled">
    <Button variant="outline" disabled>
      test
    </Button>
  </Story>
  <Story name="outline-icon-1">
    <Button variant="outline" size="Small" gapSize="md">
      <span>ファイルの編集</span>
      <TestIcon />
    </Button>
  </Story>
  <Story name="outline-round">
    <Button variant="outline" size="sm" isRound>
      リンクをコピー
    </Button>
  </Story>
  <Story name="outline-hover-fill">
    <Button variant="outline" size="sm" hoverVariant="filled" isRound>
      リンクをコピー
    </Button>
  </Story>
  <Story name="outline-with-icon-layout-left">
    <Button variant="outline" gapSize="md" withLeftIconLayoutSize="md">
      <TestIcon />
      <span>リンクをコピー</span>
    </Button>
  </Story>
  <Story name="outline-with-icon-layout-right">
    <Button variant="outline" gapSize="md" withRightIconLayoutSize="md">
      <span>リンクをコピー</span>
      <TestIcon />
    </Button>
  </Story>
</Canvas>

<ArgsTable of={Button} />
