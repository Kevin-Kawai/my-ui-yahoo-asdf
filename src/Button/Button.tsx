import * as React from "react";
import { clsx } from "clsx";
import "./Button.scss";
import { ButtonSpinner } from "./ButtonSpinner";

export type BuiButtonVariantTypes = "filled" | "outline";
export type BuiButtonSizeTypes = "xs" | "sm" | "md" | "lg" | "xl" | "2xl";
export type BuiButtonGapSizeTypes = "sm" | "md" | "lg";
export type BuiButtonIconLayoutSizeTypes = "md";

export type OwnProps = {
  /** Buttonのテーマ */
  variant?: BuiButtonVariantTypes;
  /** ホバーした時のボタンのテーマ、デフォルトは指定されたvariantによる */
  hoverVariant?: BuiButtonVariantTypes | "transparent";
  /** Buttonのサイズ、主にpaddingとheight */
  size?: BuiButtonSizeTypes;
  /** Button内のgap。Buttonは基本的にinline-flex。主にアイコンを配置した時などのため */
  gapSize?: BuiButtonGapSizeTypes;
  /** 左側にアイコンを置いたときにpaddingが同じだと間伸びして見えるため、padding-leftとpadding-rightの調整をする */
  withLeftIconLayoutSize?: BuiButtonIconLayoutSizeTypes;
  /** 右側にアイコンを置いたときにpaddingが同じだと間伸びして見えるため、padding-leftとpadding-rightの調整をする */
  withRightIconLayoutSize?: BuiButtonIconLayoutSizeTypes;
  isBlock?: boolean;
  isRound?: boolean;
  isFullWidth?: boolean;
  isLoading?: boolean;
};
type Props = OwnProps &
  React.ComponentProps<"button"> & {
    children: React.ReactNode;
  };

export const Button = ({
  children,
  className,
  variant,
  hoverVariant,
  size,
  isRound,
  isBlock,
  isFullWidth,
  withLeftIconLayoutSize,
  withRightIconLayoutSize,
  gapSize,
  isLoading,
  ...restProps
}: Props) => {
  const meregedClasses = clsx(
    "buiButtonBase",
    {
      buiButtonVariantOutline: variant === "outline",
      buiButtonVariantFilled: variant === "filled" || !variant,
      buiButtonHoverVariantOutline: hoverVariant === "outline",
      buiButtonHoverVariantFilled: hoverVariant === "filled",
      buiButtonHoverVariantTransparent: hoverVariant === "transparent",
      buiButtonSizeXs: size === "xs",
      buiButtonSizeSm: size === "sm",
      buiButtonSizeMd: size === "md",
      buiButtonSizeLg: size === "lg",
      buiButtonSizeXl: size === "xl",
      buiButtonSize2Xl: size === "2xl",
      buiButtonBlock: isBlock,
      buiButtonRound: isRound,
      buiButtonFullWidth: isFullWidth,
      buiButtonGapSizeMd: gapSize === "md",
      buiButtonLeftIconLayoutMd: withLeftIconLayoutSize === "md",
      buiButtonRightIconLayoutMd: withRightIconLayoutSize === "md",
    },
    className
  );
  return (
    <button className={meregedClasses} {...restProps}>
      {isLoading ? <ButtonSpinner /> : children}
    </button>
  );
};
