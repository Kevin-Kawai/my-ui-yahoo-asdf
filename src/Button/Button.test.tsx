import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Button } from "./Button";

test("Button", () => {
  const { container: c1 } = render(<Button>default</Button>);
  expect(c1).toMatchSnapshot();

  const { container: c2 } = render(
    <Button variant="filled" hoverVariant="outline" size="lg" gapSize="sm">
      sm
    </Button>
  );
  expect(c2).toMatchSnapshot();

  const { container: c3 } = render(
    <Button
      variant="outline"
      gapSize="md"
      hoverVariant="filled"
      withLeftIconLayoutSize="md"
      isRound
    >
      sm
    </Button>
  );
  expect(c3).toMatchSnapshot();

  const { container: c4 } = render(<Button isBlock>sm</Button>);
  expect(c4).toMatchSnapshot();
});
