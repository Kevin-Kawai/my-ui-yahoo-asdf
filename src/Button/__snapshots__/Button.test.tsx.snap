// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button 1`] = `
<div>
  <button
    class="buiButtonBase buiButtonVariantFilled"
  >
    default
  </button>
</div>
`;

exports[`Button 2`] = `
<div>
  <button
    class="buiButtonBase buiButtonVariantFilled buiButtonHoverVariantOutline buiButtonSizeLg"
  >
    sm
  </button>
</div>
`;

exports[`Button 3`] = `
<div>
  <button
    class="buiButtonBase buiButtonVariantOutline buiButtonHoverVariantFilled buiButtonRound buiButtonGapSizeMd buiButtonLeftIconLayoutMd"
  >
    sm
  </button>
</div>
`;

exports[`Button 4`] = `
<div>
  <button
    class="buiButtonBase buiButtonVariantFilled buiButtonBlock"
  >
    sm
  </button>
</div>
`;
